### ISSUE TITLE
Implement "themed" groups of rooms in room module

### ISSUE THREAD

**Issue by CCalderon01** [uchicago-cs/chiventure#540]
Research the attribute documentation from game-state and see how it will be applied to items that will be used for exit/entry (i.e. doors). Implement  a fixed "valid_exit" attribute to items placed in rooms as exits. If there's time, hard-code other attributes to be attached to different items.

**Comment by CCalderon01** [comment]
Commit: 5f2df53
I copied the attribute structure found in game-state, but made some changes to the functions. I wrote an attribute_init, attribute_new, attribute_free, attribute_add_to_hash, set_attribute_int, and set_exit_attribute functions. Now, attributes specified by a string name and data type can be assigned to items. I explicitly specified the exit attribute because I want this attribute to be assigned to every item. However, only an item of type DOOR will have their exit attribute set to 'True' (that is, 1). All other items will have a 'False' (or 0) value for the exit attribute. This is so that in future sprints we can have a item-based path implementation. That is, only items with exit attributes set to 'True' can be acted upon to connect to another room. 
In previous implementations of chiventure, paths were based on direction (north, south, east, west). It is my opinion that having to act on objects (like doors) is more intuitive and more interesting.

In summary:
- Implemented fixed attribute for items module
- Still need to write tests to check this functionality before closing this issue.

**Comment by CCalderon01** [comment]
Upon further consultation with my senior developer, we are shifting the focus of this issue to instead hardcode several "themed" groups of rooms. This will be used in the path generation module and by the autogeneration module.

**Comment by CCalderon01** [comment]
Commit: 517603d73
I created three "themed" room groupings: school, farmhouse, and castle. I also added several new fixed definition room types (ex. closet, classroom, cafeteria) that correspond to at least one of the room groupings. I created several more fixed item definitions (26 total). I implemented specific "add item to room" functions depending on the room type. For a specific room, there are specific subset items that correspond, and depending on how many items you want to be added to a room, I'm using the C library rand() function to "randomly" iterate and assign these items to a room. Finally I renamed the room and item structs to fixed_room and fixed_item to distinguish from the game-state structs.

**Comment by CCalderon01** [comment]
Commit: 65113b9 in openworld/sample-generation branch
In this commit, I wrote tests for sample_rooms and sample_items modules. They all pass! This means that now we can initialize items/rooms, create new items/rooms (from their fixed descriptions), free items/rooms, add items to rooms (with or without knowing the room type), and generate a random room type with a random number of items. ("random" meaning that it's one of the 11 rooms and up to 5 items of the 26 different types ... repetition of items is allowed).
Now that all the tests pass, I'm looking to create a pull request and close this issue soon.

**Comment by CCalderon01** [comment]
Commit 738e3f0 on openworld/sample-generation branch fixed sample_items module tests.
Commit f68d407d5 on openworld/sample-generation branch fixed sample_rooms module tests.

From my senior developer's feedback, I integrated my fixed_room and fixed_item structs with the room and item structs already found in the game-state header files. This means that I'm not trying to reinvent the structs that were already made, but I'm just adding an enumerated tag to distinguish the "fixed" struct versus the "regular" one. I had to modify the tests accordingly. I also condensed all "add item to x" functions in sample_rooms into one master function that takes in  a fixed_room definition and number of items wanted.

**Comment by CCalderon01** [comment]
I restructured the items and rooms module (again) so that I'm not repeating the room.h or item.h structures found in game-state. I am not introducing any new structs in the room and items module, but I am pulling from the structs in game-state/item.h, game-state/room.h, openworld/gen-structs.h.
The central functions in this update are the make_default_items() and make_default_room():
- make_default_items(room name, linked list of items): returns the subset of defined default items that are allowed for a given room name. If the room name doesn't have any default approved items, return the original item list.  However, if the room name does have approved items, then add these additional items to whatever is in llist. If llist is NULL, then you'll return only the list of approved items. If for some reason you want to add additional items to the allowed items list, then you would pass these through the linked list.
- make_default_room(bucket, OPTIONAL short_desc, OPTIONAL long_desc, list of allowed_items) adds all the defined default rooms for a themed "bucket" (a name for a group of related rooms). In the case where the "bucket" IS NOT part of the default definitions, there are optional parameters for a short_desc, long_desc that will be used to define a new roomspec. (Optional meaning that there won't be any error thrown if they're empty). In the case where "bucket" IS part of the default definitions (i.e. "bucket" is "school", "farmhouse", "castle"), then add all the related rooms and all their related items to the roomspec hash.

**Comment by CCalderon01** [comment]
Pull request started here: https://github.com/uchicago-cs/chiventure/pull/631
I moved my work onto a new branch, openworld/default, and compressed my default_items and default_rooms module into one big function in each. make_default_items() now doesn't take any input but returns a hash with all the default items added. Then, in make_default_room(), there is space for optional strings and allowed_items so that in case a default room/set of items isn't defined, then you can create a unique roomspec. Otherwise, if there is a match, then select items will be added to the allowed_items list through multiple calls to the default items hash. Finally, these roomspecs will be added to a larger hash and returned.
In the future, npcs will be added to the roomspec struct and thus will be another parameter in make_default_room. I'm thinking that the npc module will have their own "make_default_npc" function similar to make_default_items, and then we'll call on both of those when we make the default room.

**Comment by CCalderon01** [comment]
The pull request was merged succesfully into the dev branch.

As a final summary of this task, here are some important points:
- There are two modules (default_room, default_item) added to openworld branch that were merged onto the dev branch. In future sprints, we can look forward to integrating these modules with openworld's autogeneration module and the rest of the chiventure features.
- The default_item module hard codes 20+ items into an item hash table for later reference by default_rooms. We made this a hash table because this type of structure is flexible if we want to add more items later.
- The default_room module loads the default_item hash table and creates a roomspec hash table to store all the allowed room types and respective item types given a theme (school, farmhouse, castle, other).

**Comment by dwahme** [comment]
Issue Score: ✔️+

Comments:
Great job! Here are a few things you may want to improve in future issues:

The summary for this issue is too terse. A good rule of thumb is that you should be writing at least one full paragraph in your summary (see https://uchicago-cs.github.io/cmsc22000/project/policies/#a-descriptive-summary for examples). If this issue was generated from a Scrum board card, please remember that the summary you write in the Sprint meeting is meant to be a quick draft; it needs to be fleshed out when you convert the card to an issue.
